File: src/ui/web/backend/api/notifications.py
"""
Notifications API Routes

Provides notification endpoints.
"""

import logging

from fastapi import APIRouter, HTTPException, Depends, Query
from pydantic import BaseModel

from api.auth import get_current_user
from gateway.providers.hub import get_data_provider
from gateway.providers.data.models import (
    NotificationDTO,
    PaginatedResponse,
)

logger = logging.getLogger(__name__)
router = APIRouter(prefix="/notifications", tags=["Notifications"])


# =============================================================================
# Response Models
# =============================================================================


class UnreadCountResponse(BaseModel):
    """Unread count response"""
    count: int


class MarkAllReadResponse(BaseModel):
    """Mark all read response"""
    ok: bool
    count: int


# =============================================================================
# Endpoints
# =============================================================================


@router.get("/", response_model=PaginatedResponse)
async def list_notifications(
    unread_only: bool = Query(False),
    page: int = Query(1, ge=1),
    page_size: int = Query(20, ge=1, le=100),
    current_user: dict = Depends(get_current_user),
):
    """List user's notifications"""
    provider = get_data_provider()
    result = await provider.notifications.list_notifications(
        user_id=current_user["id"],
        unread_only=unread_only,
        page=page,
        page_size=page_size,
    )
    return result


@router.get("/unread-count", response_model=UnreadCountResponse)
async def get_unread_count(
    current_user: dict = Depends(get_current_user),
):
    """Get unread notification count"""
    provider = get_data_provider()
    count = await provider.notifications.get_unread_count(current_user["id"])
    return UnreadCountResponse(count=count)


@router.get("/{notification_id}")
async def get_notification(
    notification_id: str,
    current_user: dict = Depends(get_current_user),
):
    """Get single notification by ID"""
    provider = get_data_provider()
    notification = await provider.notifications.get_notification(
        user_id=current_user["id"],
        notification_id=notification_id,
    )

    if not notification:
        raise HTTPException(status_code=404, detail="Notification not found")

    return {"ok": True, "notification": notification}


@router.patch("/{notification_id}/read")
async def mark_as_read(
    notification_id: str,
    current_user: dict = Depends(get_current_user),
):
    """Mark notification as read"""
    provider = get_data_provider()
    success = await provider.notifications.mark_as_read(
        user_id=current_user["id"],
        notification_id=notification_id,
    )

    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")

    return {"ok": True, "message": "Notification marked as read"}


@router.post("/read-all", response_model=MarkAllReadResponse)
async def mark_all_as_read(
    current_user: dict = Depends(get_current_user),
):
    """Mark all notifications as read"""
    provider = get_data_provider()
    count = await provider.notifications.mark_all_as_read(current_user["id"])
    return MarkAllReadResponse(ok=True, count=count)


@router.delete("/{notification_id}")
async def delete_notification(
    notification_id: str,
    current_user: dict = Depends(get_current_user),
):
    """Delete notification"""
    provider = get_data_provider()
    success = await provider.notifications.delete_notification(
        user_id=current_user["id"],
        notification_id=notification_id,
    )

    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")

    return {"ok": True, "message": "Notification deleted"}

File: src/ui/web/frontend/src/api/notifications.js
/**
 * Notifications API - Gateway API Access
 * Single responsibility: Notification CRUD operations
 */

import { get, patch, post, del } from '@/api/client'
import { ENDPOINTS } from '@/api/config'
import { DEFAULTS } from '@/config/defaults'

/**
 * Get user notifications
 * @param {Object} params - Query parameters
 * @param {number} params.page - Page number
 * @param {number} params.page_size - Items per page
 * @param {boolean} params.unread_only - Only unread notifications
 * @returns {Promise<{ok: boolean, items: Array, total: number}>}
 */
export async function getNotifications(params = {}) {
  try {
    const { page = 1, page_size = DEFAULTS.PAGINATION.DEFAULT, unread_only = false } = params

    const result = await get(ENDPOINTS.NOTIFICATIONS.LIST, {
      params: {
        page,
        page_size,
        unread_only
      }
    })

    return {
      ok: true,
      items: result.items || [],
      total: result.total || 0
    }
  } catch (err) {
    return { ok: false, error: err.userMessage || err.message, items: [], total: 0 }
  }
}

/**
 * Get unread notification count
 * @returns {Promise<{ok: boolean, count: number}>}
 */
export async function getUnreadCount() {
  try {
    const result = await get(ENDPOINTS.NOTIFICATIONS.UNREAD_COUNT)
    return { ok: true, count: result.count || 0 }
  } catch (err) {
    return { ok: false, error: err.userMessage || err.message, count: 0 }
  }
}

/**
 * Mark notification as read
 * @param {string} notificationId - Notification ID
 * @returns {Promise<{ok: boolean}>}
 */
export async function markAsRead(notificationId) {
  try {
    await patch(ENDPOINTS.NOTIFICATIONS.MARK_READ(notificationId))
    return { ok: true }
  } catch (err) {
    return { ok: false, error: err.userMessage || err.message }
  }
}

/**
 * Mark all notifications as read
 * @returns {Promise<{ok: boolean, count: number}>}
 */
export async function markAllAsRead() {
  try {
    const result = await post(ENDPOINTS.NOTIFICATIONS.MARK_ALL_READ)
    return { ok: true, count: result.count || 0 }
  } catch (err) {
    return { ok: false, error: err.userMessage || err.message, count: 0 }
  }
}

/**
 * Delete a notification
 * @param {string} notificationId - Notification ID
 * @returns {Promise<{ok: boolean}>}
 */
export async function deleteNotification(notificationId) {
  try {
    await del(ENDPOINTS.NOTIFICATIONS.GET(notificationId))
    return { ok: true }
  } catch (err) {
    return { ok: false, error: err.userMessage || err.message }
  }
}

export default {
  getNotifications,
  getUnreadCount,
  markAsRead,
  markAllAsRead,
  deleteNotification
}
